File: code/packages/lua/type_checker_protocol/tests/test_type_checker_protocol.lua
package.path = "../src/?.lua;" .. "../src/?/init.lua;" .. package.path

local protocol = require("coding_adventures.type_checker_protocol")

describe("type_checker_protocol", function()
    it("normalizes hook kinds during dispatch", function()
        local checker = protocol.new_generic_type_checker(function(node)
            return node.kind
        end)

        checker:register_hook("enter", "fn decl", function()
            return "exact"
        end)

        assert.equals("exact", checker:dispatch("enter", { kind = "fn decl" }))
    end)

    it("falls through when a hook returns not_handled", function()
        local checker = protocol.new_generic_type_checker(function(node)
            return node.kind
        end)

        checker:register_hook("enter", "expr:add", function()
            return checker:not_handled()
        end)
        checker:register_hook("enter", "*", function()
            return "fallback"
        end)

        assert.equals("fallback", checker:dispatch("enter", { kind = "expr:add" }))
    end)

    it("collects diagnostics into a type-check result", function()
        local checker = protocol.new_generic_type_checker(
            function(node)
                return node.kind
            end,
            function(node)
                return node.line, node.column
            end
        )

        function checker:run(ast)
            self:error("bad node", ast)
        end

        local result = checker:check({ kind = "expr", line = 4, column = 2 })

        assert.is_false(result.ok)
        assert.equals(1, #result.errors)
        assert.equals("bad node", result.errors[1].message)
        assert.equals(4, result.errors[1].line)
        assert.equals(2, result.errors[1].column)
    end)
end)

File: code/packages/ruby/type_checker_protocol/test/test_type_checker_protocol.rb
# frozen_string_literal: true

require_relative "test_helper"

class DummyChecker < CodingAdventures::TypeCheckerProtocol::GenericTypeChecker
  def initialize
    super(
      node_kind: ->(node) { node[:kind] },
      locate: ->(node) { [node.fetch(:line, 1), node.fetch(:column, 1)] }
    )
  end

  def run(ast)
    dispatch("enter", ast)
  end
end

class TypeCheckerProtocolTest < Minitest::Test
  def test_exact_hook_dispatches
    checker = DummyChecker.new
    checker.register_hook("enter", "fn decl") { |_node| :exact }

    assert_equal :exact, checker.dispatch("enter", { kind: "fn decl" })
  end

  def test_wildcard_hook_dispatches_when_exact_missing
    checker = DummyChecker.new
    checker.register_hook("enter", "*") { |_node| :wildcard }

    assert_equal :wildcard, checker.dispatch("enter", { kind: "let stmt" })
  end

  def test_not_handled_falls_through_to_wildcard
    checker = DummyChecker.new
    checker.register_hook("enter", "expr:add") { |_node| checker.not_handled }
    checker.register_hook("enter", "*") { |_node| :fallback }

    assert_equal :fallback, checker.dispatch("enter", { kind: "expr:add" })
  end

  def test_check_returns_result_with_errors
    checker = DummyChecker.new
    checker.register_hook("enter", "*") do |node|
      checker.error("bad node", node)
    end

    result = checker.check({ kind: "expr:add", line: 3, column: 7 })

    assert_equal false, result.ok
    assert_equal 1, result.errors.length
    assert_equal "bad node", result.errors.first.message
    assert_equal 3, result.errors.first.line
    assert_equal 7, result.errors.first.column
  end
end
